### DISCUSSION TITLE
File information inconsistencies and GetVolumeInfo update frequency in memfs

### DISCUSSION THREAD

**Discussion by Ceiridge** [winfsp/winfsp#417] [Q&A]
When using WinFsp with **memfs**, the file information dialog of the Windows Explorer is different:
![winfsp-file-info](https://user-images.githubusercontent.com/40047520/162271210-a4a65b20-311f-41db-9627-ac47ae073a85.png)

### File flags
One difference is that with WinFsp, you can **only set the archive flag** _(asides from the hidden and read-only flags)_ of the file (which then is always re-set to 1 anyway: https://github.com/winfsp/winfsp/blob/master/tst/memfs/memfs.cpp#L1325), whereas on an **NTFS** drive, you can open another dialog with **more attributes**, including compression and encryption.\
Memfs actually has file _"encryption"_ with ROT13: https://github.com/winfsp/winfsp/blob/master/tst/memfs/memfs.cpp#L2156

This makes me wonder how the developer managed to test it without noticing this inconsistency. If I remember correctly, this can also happen to drives that are not formatted with NTFS or are USB/removable drives.\
The file system's name of memfs is set to *NTFS*:
![memfs](https://user-images.githubusercontent.com/40047520/162272973-67fd59cf-3a1f-4d3c-bd06-e3621344726f.png)

### AllocationSize caching?
Also, the file's size on disk/**AllocationSize** never changes after opening the file information dialog (in this case: 512 bytes allocated, although more than 1 KB is written). I also verified the AllocationSize with a debugger and although the right size is returned, Windows shows an old one. In comparison to this, the actual file size is always correctly displayed. This issue **does not occur on NTFS drives**.

### GetVolumeInfo update frequency
Another strange thing I observed is that _GetVolumeInfo_ is only called when a file is created/renamed/modified, but not when you reload the view of all drives in Windows Explorer. Is there a way to **update the volume info without** requiring such **user interaction**?

**Comment by billziss-gh** [comment]
@Ceiridge 

### File flags

WinFsp/MEMFS does not support all NTFS features. Please see the full list of [unsupported features](https://github.com/winfsp/winfsp/wiki/NTFS-Compatibility#unsupported-features)

For compressed and encrypted files specifically:

- Compressed files. A user mode file system is free to implement compressed files. However WinFsp does not support any compression related FSCTL or information class codes.

- Encrypted files. A user mode file system is free to implement encrypted files. However WinFsp does not support any encryption related FSCTL or information class codes.

### AllocationSize caching?

WinFsp/MEMFS reports the right size (otherwise things would not work), so I am not sure why the discrepancy. Did you mount MEMFS as a disk or network file system?

### GetVolumeInfo update frequency

`GetVolumeInfo` is only called when an application or the system calls it. There is no mechanism to notify of volume info changes and force the system to call `GetVolumeInfo`.

There is however a directory change notification mechanism that may or may not force the system to call `GetVolumeInfo` to get the new size.

**Reply by Ceiridge** [reply]
I was wondering why the Windows Explorer only shows the option to set the archive file flag and not the other flags, even if they are not implemented by the user mode file system. Normally, there is a **button called `Advanced...`.**\
 I just tested it with a removable drive formatted with FAT, exFAT, NTFS and all of these have that button, but not MEMFS. I also made sure that MEMFS reports NTFS as its file system name.\
I believe that **during the creation** of a file system (maybe in user space, maybe in the WinFsp driver?), some value is different to these drives that makes the Windows Explorer behave differently.

The discrepancy of the **AllocationSize** can also only be seen in the Windows Explorer. It does not happen with normal drives (like the ones mentioned above) and to reproduce it, do the following:
1. Create a new file and write & save some stuff in it
2. Open the file properties dialog and look at the file size on disk (AllocationSize). Then close the dialog
3. Write more to the file or delete some stuff until the AllocationSize changes
4. Check the file properties dialog again and see that the **AllocationSize has remained the same**, even if for example the actual file size is much bigger

_Note: At least the user space file system reports the right AllocationSize, as I have checked that with a debugger._

Both issues occur on MEMFS mounted as a disk and network file system.

So it would be possible to **notify** the system **about a new file**, for example in the root directory, with `FspFileSystemNotify` and then Windows potentially calls `GetVolumeInfo` at some point because of it?\

**Reply by Ceiridge** [reply]
@billziss-gh
